### PR TITLE
imporve implementation and doc of roles

### PR DESCRIPTION
**Pull request by MajdT51**

### CODE REVIEW

**Review by matthiashub-porsche — CHANGES_REQUESTED** [review]

**Review by jpbu — COMMENTED** [review]
Reviewed it in GitHub, looks good, a few comments.

File: src/main/kotlin/org/eclipse/tractusx/managedidentitywallets/routes/AuthorizationHandler.kt [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        call: ApplicationCall,
+        identifier: String?,
+        role: String
+    ): AuthorizationResponse {
+        val principal = AuthConstants.getPrincipal(call.attributes)
+        if (principal == null || principal.role.isBlank()) {
+            return AuthorizationResponse(false, "Authorization failed: Principal is null " +
+                    "or it has an empty role")
+        }
+        if (principal.role == role) {
```
[the remark this answers is not in the document]
Comment (matthiashub-porsche) [reply src/main/kotlin/org/eclipse/tractusx/managedidentitywallets/routes/AuthorizationHandler.kt:72]: To my understanding this function is called with a single role only (done by the _notarized_ thing around)

[the remark this answers is not in the document]
Comment (jpbu) [reply src/main/kotlin/org/eclipse/tractusx/managedidentitywallets/routes/AuthorizationHandler.kt:72]: We should validate that with the local Keycloak setup if not yet done. In the ApplicationTest it looks like only JWTs with a single role are created, does that correctly reflect how they are used in the Keycloak scenario?

[the remark this answers is not in the document]
Comment (matthiashub-porsche) [reply src/main/kotlin/org/eclipse/tractusx/managedidentitywallets/routes/AuthorizationHandler.kt:72]: Yes, great suggestion to put it in the test case, full ack - I only tested with Postman my earlier implementation

[the remark this answers is not in the document]
Comment (matthiashub-porsche) [reply src/main/kotlin/org/eclipse/tractusx/managedidentitywallets/routes/AuthorizationHandler.kt:72]: Here the keycloak example:

```
  "resource_access": {
    "Cl5-CX-Custodian": {
      "roles": [
        "delete_wallets",
        "update_wallets",
        "view_wallets",
        "add_wallets"
      ]
    }
  },
```

File: src/test/kotlin/org/eclipse/tractusx/managedidentitywallets/ApplicationTest.kt [hunk 2]
```diff
@@ -354,29 +354,41 @@ class ApplicationTest {
             Services.businessPartnerDataService = bpdService
         }) {
             // view wallets with wrong token should not work
-            handleRequest(HttpMethod.Get, "/api/wallets") {
-                addHeader(HttpHeaders.Authorization, "Bearer $UPDATE_TOKEN")
-                addHeader(HttpHeaders.Accept, ContentType.Application.Json.toString())
-            }.apply {
-                assertEquals(HttpStatusCode.Unauthorized, response.status())
+            var exception = assertFailsWith<ForbiddenException> {
+                handleRequest(HttpMethod.Get, "/api/wallets") {
```
Comment (matthiashub-porsche) [inline src/test/kotlin/org/eclipse/tractusx/managedidentitywallets/ApplicationTest.kt:358]: Is the return code of `403` returned in the HTTP call? Could we also add an additional test for that?

**Comment by matthiashub-porsche** [comment]
> Do we still need the roleMap and roleMappings? It looks like it is now only used in a test

@jpbu Quick follow-up question: where do you see that? For me the configuration check on startup is quite valuable:

```
        if (roleMap[AuthorizationHandler.ROLE_CREATE_WALLETS] == null) {
             log.error("Configuration error, ${AuthorizationHandler.ROLE_CREATE_WALLETS} role mapping not defined, system will not behave correctly!")
             throw Exception("Configuration error, ${AuthorizationHandler.ROLE_CREATE_WALLETS} role mapping not defined, system will not behave correctly!")
         }
```

**Comment by jpbu** [comment]
> > Do we still need the roleMap and roleMappings? It looks like it is now only used in a test
> 
> @jpbu Quick follow-up question: where do you see that? For me the configuration check on startup is quite valuable:
> 
> ```
>         if (roleMap[AuthorizationHandler.ROLE_CREATE_WALLETS] == null) {
>              log.error("Configuration error, ${AuthorizationHandler.ROLE_CREATE_WALLETS} role mapping not defined, system will not behave correctly!")
>              throw Exception("Configuration error, ${AuthorizationHandler.ROLE_CREATE_WALLETS} role mapping not defined, system will not behave correctly!")
>          }
> ```

@matthiashub-porsche I don't see the `roleMap` used anywhere, same as the `roleMappings` property. In my understanding it was only necessary for mapping the jwtConfigs to the roles, this has now been removed and the roles are checked in the `AuthorizationHandler`.  I quickly tried to remove the `roleMap` from the code, all compiles successfully and besides the `testMissingRoleMapping` test case that explicitly (and only) tests this role mapping all the other test cases run successfully.

**Comment by matthiashub-porsche** [comment]
> > > Do we still need the roleMap and roleMappings? It looks like it is now only used in a test
> > 
> > 
> > @jpbu Quick follow-up question: where do you see that? For me the configuration check on startup is quite valuable:
> > ```
> >         if (roleMap[AuthorizationHandler.ROLE_CREATE_WALLETS] == null) {
> >              log.error("Configuration error, ${AuthorizationHandler.ROLE_CREATE_WALLETS} role mapping not defined, system will not behave correctly!")
> >              throw Exception("Configuration error, ${AuthorizationHandler.ROLE_CREATE_WALLETS} role mapping not defined, system will not behave correctly!")
> >          }
> > ```
> 
> @matthiashub-porsche I don't see the `roleMap` used anywhere, same as the `roleMappings` property. In my understanding it was only necessary for mapping the jwtConfigs to the roles, this has now been removed and the roles are checked in the `AuthorizationHandler`. I quickly tried to remove the `roleMap` from the code, all compiles successfully and besides the `testMissingRoleMapping` test case that explicitly (and only) tests this role mapping all the other test cases run successfully.

@jpbu then it is a wrong logic =) because the roles are not named necessarily the same as we name it internally, e.g. the portal named it `add_wallets` instead of `create_wallets`, thus I added the map lookup. I've pushed a small update, maybe @MajdT51 can review and test on his end.

File: src/main/kotlin/org/eclipse/tractusx/managedidentitywallets/models/BusinessPartnerDataDto.kt [hunk 3]
```diff
@@ -50,6 +52,13 @@ data class BPDMConfig(
     val scope: String
 )
 
+@Serializable
+data class BusinessPartnerDataDtoParameter(
```
[the remark this answers is not in the document]
Comment (MajdT51) [reply src/main/kotlin/org/eclipse/tractusx/managedidentitywallets/models/BusinessPartnerDataDto.kt]: done

File: src/main/kotlin/org/eclipse/tractusx/managedidentitywallets/models/BusinessPartnerDataDto.kt [hunk 4]
```diff
@@ -50,6 +52,13 @@ data class BPDMConfig(
...
+    @Param(type = ParamType.QUERY)
+    @Field(description = "The DID or BPN of the Wallet to update", name = "identifier")
```
[the remark this answers is not in the document]
Comment (MajdT51) [reply src/main/kotlin/org/eclipse/tractusx/managedidentitywallets/models/BusinessPartnerDataDto.kt]: done

File: src/main/kotlin/org/eclipse/tractusx/managedidentitywallets/routes/AuthorizationHandler.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (MajdT51) [reply src/main/kotlin/org/eclipse/tractusx/managedidentitywallets/routes/AuthorizationHandler.kt]: done

File: src/main/kotlin/org/eclipse/tractusx/managedidentitywallets/routes/AuthorizationHandler.kt [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    private fun checkIfBpnMatchesToPrincipalBpn(
+        givenBpn: String,
+        principalBpn: String?
+    ) : AuthorizationResponse {
+        return if (!principalBpn.isNullOrEmpty() && givenBpn == principalBpn) {
+            log.debug("Authorization successful: Wallet BPN $givenBpn does match requestors BPN $principalBpn")
+            AuthorizationResponse(true)
+        } else {
+            val errorMsg = "Authorization failed: Wallet BPN $givenBpn does not match requestors BPN $principalBpn"
+            log.error(errorMsg)
```
[the remark this answers is not in the document]
Comment (MajdT51) [reply src/main/kotlin/org/eclipse/tractusx/managedidentitywallets/routes/AuthorizationHandler.kt]: outdated

File: src/test/kotlin/org/eclipse/tractusx/managedidentitywallets/ApplicationTest.kt (hunk 2 above)
Comment (MajdT51) [reply src/test/kotlin/org/eclipse/tractusx/managedidentitywallets/ApplicationTest.kt]: done. added mapping error -> http code to the testcases

**Review by jpbu — APPROVED** [review]
Double-checked and tested locally with the latest version, looks good to me now.

**Pull request merged**